### PR TITLE
Support lambda functions (ex: `aggregate`) and add several missing Spark functions

### PR DESCRIPTION
**Pull request by shangyian**
### Summary

This change adds the following Spark functions:
* `element_at`
* `split`
* `aggregate`
* `first`
* `cardinality`

Note that `aggregate` is a bit of a special case function that relates to https://github.com/DataJunction/dj/issues/543, since it takes in a lambda function as an argument. In order to support other functions that take lambdas (ex: `map_zip_with`, `map_filter`), we modify the way that `ast.Function` is compiled:
1. Compile the arguments of the function
2. Compile the lambda function expression. This is done with a custom method called `compile_lambda` that is defined on each such function's class. The main purpose of this function is to add types to the input arguments of the lambda expression so that we can infer the return type of the lambda by evaluating the expression.
3. Compile all other child elements of the function, if any.

### Test Plan

Added some unit tests. Tested with DJ server locally for a complex query with all of these functions used.

- [X] PR has an associated issue: #543, #544
- [x] `make check` passes
- [x] `make test` shows 100% unit test coverage

### Deployment Plan

N/A

### CODE REVIEW

File: tests/sql/functions_test.py [hunk 1]
```diff
@@ -256,3 +258,70 @@ def test_floor(types, expected) -> None:
...
+        when acc = '' then element_at(split(x, '::'), 1)
+        when acc = 'a' then acc
+        else element_at(split(x, '::'), 1) end)) as item
+    from (
+      select 1 as id, ARRAY('b', 'c', 'a', 'x', 'g', 'z') AS items
+    )
+    """,
+    )
+    exc = DJException()
+    ctx = ast.CompileContext(session=session, exception=exc)
```
Comment (samredai) [inline tests/sql/functions_test.py:280]: Should these tests also include an assert that the `exc` variable doesn't contain any errors?

**Review by samredai — APPROVED** [review]
LGTM, just one comment about the tests.

**Review by agorajek — APPROVED** [review]
Nice one @shangyian !

File: tests/sql/functions_test.py (hunk 1 above)
Comment (shangyian) [reply tests/sql/functions_test.py:280]: Good point, added!

**Pull request merged**